### PR TITLE
cmd/tailscale/cli: do not allow turning Funnel on while shields-up

### PR DESCRIPTION
**Pull request by shayne**
Fixes: tailscale/corp#9027

### CODE REVIEW

File: cmd/tailscale/cli/funnel.go [hunk 1]
```diff
@@ -73,6 +73,18 @@ func (e *serveEnv) runFunnel(ctx context.Context, args []string) error {
...
+	prefs, err := e.lc.GetPrefs(ctx)
+	if err != nil {
+		return fmt.Errorf("getting prefs: %w", err)
+	}
+
+	if on && prefs.ShieldsUp {
+		return fmt.Errorf("%s\n%s",
+			"Unable to turn on Funnel while shields-up=true.",
+			"Run `tailscale set --shields-up=false` to turn off Shields.")
+	}
```
Comment (maisem) [inline cmd/tailscale/cli/funnel.go:86]: I think you want to do this in local.go
https://github.com/tailscale/tailscale/blob/f4da995940ea13bda69979ec1a07251c6b757c5d/ipn/ipnlocal/local.go#L2532

File: ipn/ipnlocal/serve.go [hunk 2]
```diff
@@ -218,6 +218,11 @@ func (b *LocalBackend) SetServeConfig(config *ipn.ServeConfig) error {
 	b.mu.Lock()
 	defer b.mu.Unlock()
 
+	prefs := b.sanitizedPrefsLocked()
```
Comment (maisem) [inline ipn/ipnlocal/serve.go:221]: you can do `prefs := b.pm.CurrentPrefs()`

File: ipn/ipnlocal/serve.go [hunk 3]
```diff
@@ -218,6 +218,11 @@ func (b *LocalBackend) SetServeConfig(config *ipn.ServeConfig) error {
...
+	if config.IsFunnelOn() && prefs.ShieldsUp() {
+		return errors.New("Unable to turn on Funnel while shields-up is enabled")
```
Comment (maisem) [inline ipn/ipnlocal/serve.go:223]: this is fine, but you also need the opposite side

**Review by maisem — APPROVED** [review]

File: ipn/ipnlocal/local.go [hunk 4]
```diff
@@ -2628,6 +2631,18 @@ func (b *LocalBackend) checkExitNodePrefsLocked(p *ipn.Prefs) error {
 	return nil
 }
 
+func (b *LocalBackend) checkFunnelEnabledLocked(p *ipn.Prefs) error {
+	sc0 := b.serveConfig
+	if !sc0.Valid() {
+		return nil
+	}
+	sc1 := sc0.AsStruct()
```
Comment (maisem) [inline ipn/ipnlocal/local.go:2639]: Oh thats an expensive call just to check if funnel is enabled. You can selectively expose methods from structs to their corresponding views, so we should add a method "IsFunnelOn" to the view and use it here.

See tailcfg.Node.IsTagged as an example.

**Review by maisem — APPROVED** [review]

File: ipn/ipnlocal/local.go (hunk 4 above)
Comment (shayne) [reply ipn/ipnlocal/local.go]: Updated. PTAL. IsTagged looked pretty standard, but I think I did what we wanted based on other view methods.

Comment (shayne) [reply ipn/ipnlocal/local.go]: [posterity] Example @ `dbx.Node.IsTagged`

Updated. PTAL

**Review by maisem — APPROVED** [review]

**Pull request merged**